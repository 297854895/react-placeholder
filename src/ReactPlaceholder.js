import React from 'react';
import {omit} from 'lodash/object';
import placeholders from './placeholders';

const ReactFiller = React.createClass({

  /* eslint-disable key-spacing */
  propTypes: {
    children:          React.PropTypes.oneOfType([
                         React.PropTypes.node,
                         React.PropTypes.element
                       ]).isRequired,
    ready:             React.PropTypes.bool.isRequired,
    firstLaunchOnly:   React.PropTypes.bool,
    type:              React.PropTypes.oneOf(['text', 'media', 'textRow', 'rect', 'round']),
    rows:              React.PropTypes.number,
    color:             React.PropTypes.string,
    customPlaceholder: React.PropTypes.oneOfType([
                         React.PropTypes.node,
                         React.PropTypes.element
                       ])
  },
  /* eslint-enable key-spacing */

  getDefaultProps() {
    return {
      type: 'text',
      color: '#CDCDCD'
    };
  },

  getInitialState() {
    return {
      ready: this.props.ready
    };
  },

  isReady() {
    return this.props.firstLaunchOnly ? this.state.ready : this.props.ready;
  },

  getFiller() {
    if (this.props.customPlaceholder) {
      return this.props.customPlaceholder;
    }
  const Placeholder = placeholders[this.props.type];
    const props = omit(this.props, ['children', 'ready', 'firstLaunchOnly', 'type']);
    return <Placeholder {...props} />;
  },

  render() {
    return this.isReady() ? this.props.children : this.getFiller();
  },

  componentWillReceiveProps(nextProps) {
    if (!this.state.ready) {
      this.setState({
        ready: nextProps.ready
      });
    }
  }

});


export default ReactFiller;
