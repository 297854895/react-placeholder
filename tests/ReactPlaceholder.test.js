import React from 'react';
import ReactPlaceholder from '../src/ReactPlaceholder';
import { shallow } from 'enzyme';

describe('ReactPlaceholder', () => {

  it('renders the placeholder when the content is not ready', () => {
    const content = <div>Some content still loading...</div>
    const tree = shallow(
      <ReactPlaceholder type='text' rows={2} ready={false}>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    expect(tree.getNodes()).toMatchSnapshot();
  });

  it('renders the content when it\'s ready', () => {
    const content = <div>Some ready content</div>
    const tree = shallow(
      <ReactPlaceholder type='text' rows={2} ready={true}>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(true);
    expect(tree.getNodes()).toMatchSnapshot();
  });

});