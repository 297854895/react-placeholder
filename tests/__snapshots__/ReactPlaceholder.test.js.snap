// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactPlaceholder renders the content when it's ready 1`] = `
Array [
  <div>
    Some ready content
</div>,
]
`;

exports[`ReactPlaceholder renders the placeholder when the content is not ready 1`] = `
Array [
  <TextBlock
    color="#CDCDCD"
    rows={2}
/>,
]
`;
